src/pveclib/vec_int128_ppc.h

```diff
@@ -5387,31 +5413,41 @@ vec_popcntq (vui128_t vra)
 {
   vui64_t result;
 
-#ifdef _ARCH_PWR8
+#ifdef _ARCH_PWR9
   /*
    * Use the Vector Population Count Doubleword instruction to get
    * the count for the left and right vector halves.  Then sum across
    * the left and right counts to get the final 128-bit vector count
    * (0-128).
    */
-  vui64_t vt1;
+  vui64_t vt1, h64, l64;
   const vui64_t vzero = { 0, 0 };
 
-  vt1 = vec_popcntd ((__vector unsigned long long) vra);
-  result = (vui64_t) vec_sums ((__vector int) vt1,
-                               (__vector int) vzero);
+  vt1 = vec_popcntd ((vui64_t)  vra);
+  h64 = vec_mrgahd ((vui128_t)vzero, (vui128_t)vt1);
+  l64 = vec_mrgald ((vui128_t)vzero, (vui128_t)vt1);
+  result = vec_addudm (h64, l64);
```

Comment:
Does the comment immediately above this code need to be updated with the new implementation?

Reply:
Yes. The implementation of P8 diverges from P9. On P8 vpopcntd is cracked on P8 for 4 cycles while vpopcntw is still 2. P9 is 3 cycles for either so used vpopcntd and expanded the dword add across for more parallelism.